In [4]:
gene_sets = {
'arsenic' : {1,2,3,4,5,6,8,12},
'cadmium' : {2,12,6,4},
'copper' : {7,6,10,4,8},
'mercury': {3,2,4,5,1}
}


In [5]:
gene_sets['arsenic']

{1, 2, 3, 4, 5, 6, 8, 12}

In [6]:
def get_pairwise_scores(sets, query1, query2):
    union={}
    intersection={}
    score_matrix={}
    sets_sorted=dict(sorted(sets.items(), key=lambda x: x[0].lower()))
    for key in sets_sorted:
        score_matrix[key]={}
        for k in sets_sorted:
            if key<k:
                union[key+'+'+k]=sets_sorted[key].union(sets_sorted[k])
                intersection[key+'+'+k]=sets_sorted[key].intersection(sets_sorted[k])
                ratio=round(len(intersection[key+'+'+k])/len(union[key+'+'+k]),2)
                score_matrix[key][k]=ratio
    if query1 not in sets or query2 not in sets:
        raise ValueError('The queries are not part of the input set.')
    elif query1<query2:
        return score_matrix[query1][query2]
    elif query1>query2:
        return score_matrix[query2][query1]
    elif query1==query2:
        return score_matrix, 1
    

In [13]:
def get_highest_pair(sets):
    union={}
    intersection={}
    score_matrix={}
    sets_sorted=dict(sorted(sets.items(), key=lambda x: x[0].lower()))
    for key in sets_sorted:
        score_matrix[key]={}
        for k in sets_sorted:
            if key<k:
                union[key+'+'+k]=sets_sorted[key].union(sets_sorted[k])
                intersection[key+'+'+k]=sets_sorted[key].intersection(sets_sorted[k])
                ratio=round(len(intersection[key+'+'+k])/len(union[key+'+'+k]),2)
                score_matrix[key][k]=ratio
    maximum={}
    max_val=0
    for key in score_matrix:
        for k in score_matrix[key]:
            if score_matrix[key][k]>max_val:
                max_val=score_matrix[key][k]
                string=key+'+'+k
    maximum[string]=max_val
    return maximum

In [8]:
get_pairwise_scores(gene_sets, 'copper','arsenic')

0.3

In [9]:
get_pairwise_scores(gene_sets, 'arsenic','mercury')

0.62

In [10]:
get_pairwise_scores(gene_sets,'cadmium','cadmium')

({'arsenic': {'cadmium': 0.5, 'copper': 0.3, 'mercury': 0.62},
  'cadmium': {'copper': 0.29, 'mercury': 0.29},
  'copper': {'mercury': 0.11},
  'mercury': {}},
 1)

In [11]:
get_pairwise_scores(gene_sets,'cadmium','iron')

ValueError: The queries are not part of the input set.

In [14]:
get_highest_pair(gene_sets)

{'arsenic+mercury': 0.62}

The highest similarity score has the combination of arsenic and mercury with a score of 0.62. So the combination of mercury and arsenic affects the most genes. Copper and mercury have the lowest score with a value of 0.11. 